File: sources/app/src/contracts/TimeEntryContract.ts
import type { Branded } from "typings/brand";
import type { ActivityContract, ActivityId } from "@/contracts/ActivityContract";
import type { ProjectContract, ProjectId } from "@/contracts/ProjectContract";

export type TimeEntryCreateContract = {
    taskId: string | null;
    dateStarted: Date;
    dateEnded: Date;
    projectId: ProjectId;
    activityId: ActivityId;
    comment: string | null;
};

export type TimeEntryId = Branded<number, "timeEntryId"> | null;

export type TimeEntryContract = {
    id: TimeEntryId;
    userId: number;
    taskId: string | null;
    project: ProjectContract;
    activity: ActivityContract;
    breakDetails: TimeEntryBreakContract | null;
    dateStarted: Date;
    dateEnded: Date;
    comment: string | null;
};

export type TimeEntryBreakContract = {
    durationMillis: number;
    valid: boolean;
};

export type TimeEntryUpdateContract = {
    taskId: string | null;
    projectId: ProjectId;
    activityId: ActivityId;
    dateStarted: Date;
    dateEnded: Date;
    comment: string | null;
};

File: sources/Vuetrack.Api/Vuetrack.Backends.Abstractions/Contracts/TimeEntryContract.cs
namespace Vuetrack.Backends.Abstractions.Contracts;

public sealed record TimeEntryContract
{
    public string? Id { get; init; }

    public required string UserId { get; init; }

    public string? TaskId { get; init; }

    public required ProjectContract Project { get; init; }

    public required ActivityContract Activity { get; init; }

    public TimeEntryBreakContract? BreakDetails { get; init; }

    public required DateTime DateStarted { get; init; }

    public required DateTime DateEnded { get; init; }

    public string? Comment { get; init; }
}
